rfc/system/5103-offloaded-literal.md

```diff
@@ -4,32 +4,77 @@
+```yaml
+type LiteralOffloadingConfig struct {
+  Enabled bool
+  // Maps flytekit SDK to minimum supported version that can handle reading offloaded literals.
+  SupportedSDKVersions map[string]string
+  // Default, 10Mbs. Determines the size of a literal at which to trigger offloading
+  MinSizeInMBForOffloading uint64
+  // Fail fast threshold
+  MaxSizeInMBForOffloading uint64
+}
+
+```
 
-Questions: How will things like metadata be handled? Should they be merged? What should be in the `value` field of the main parent Literal?
+### 3.1 Offloaded Literal IDL
+Update the `Literal` [message](https://github.com/flyteorg/flyte/blob/4a7c3c0040b1995a43939407b99ca3e87b1eb752/flyteidl/protos/flyteidl/core/literals.proto#L94-L114)
+like so
+
+```protobuf
+message Literal {
+    oneof value {
+        // A simple value.
+        Scalar scalar = 1;
+        // A collection of literals to allow nesting.
+        LiteralCollection collection = 2;
+        // A map of strings to literals.
+        LiteralMap map = 3;
+    }
+    ...
+    // ** new below this line **
+    // If this literal is offloaded, this field will contain metadata including the offload location.
+    string uri = 6;
+    // Includes information about the size of the literal.
+    uint64 size_bytes = 7;
+} 
+
+```
 
 ### 3.2 Flyte Propeller
-* When writing map task outputs, depending on the size, Propeller will need to offload the LiteralCollection after constructing it, and create a new Literal for downstream tasks to use, with the 
-* Also Propeller will need to check the flytekit version of the map task. If it's an older version (i.e. before the change proposed in this RFC), and it's large enough to need to be offloaded, it should fail the task. The assumption here is that if the map task is of the older version then downstream tasks will probably also be of those older versions which won't know how to resolved these offloaded literals. 
+Once offloading is enabled in the deployment config, flytepropeller can read from the [RuntimeMetadata](https://github.com/flyteorg/flyte/blob/f448a0358d8706a09b65b96543134f629327d755/flyteidl/protos/flyteidl/core/tasks.proto#L71-L87) in the task config to determine the SDK version.
+
+When writing outputs in the [remote_file_output_writer](https://github.com/flyteorg/flyte/blob/2ca31119d6b9258661a71f38e450f93b6692402c/flyteplugins/go/tasks/pluginmachinery/ioutils/remote_file_output_writer.go#L56-L84) the source code should detect whether the literal size exceeds the configured minimum and
+- if the task is using a newer SDK version that supports reading offloaded literals, offload the literal to the configured storage backend and update the literal with the offload URI and size.
+- if the task is using an older SDK version that doesn't support offloaded literals, fail the task with an error message indicating that the task output is too large and the user should update their SDK version.
```

Comment:
let's add a line to say downstream tasks also have to upgraded?  that is, if you have a reference/remote task downstream that consumes the map task output, but it hasn't been updated, then it'll fail.